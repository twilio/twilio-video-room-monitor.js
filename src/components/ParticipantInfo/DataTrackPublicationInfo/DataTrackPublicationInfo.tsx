import React from 'react';
import { LocalDataTrack, LocalDataTrackPublication, RemoteDataTrack, RemoteDataTrackPublication } from 'twilio-video';
import useIsTrackSwitchedOff from '../../../hooks/useIsTrackSwitchedOff/useIsTrackSwitchedOff';
import useTrack from '../../../hooks/useTrack/useTrack';
import DataTrack from '../../typography/trackInformation/DataTrack/DataTrack';
import Datum from '../../typography/common/Datum/Datum';
import StatsContainer from '../../typography/common/StatsContainer/StatsContainer';

export const DataTrackInfo: React.FC<{
  track: LocalDataTrack | RemoteDataTrack;
}> = ({ track }) => {
  const isSwitchedOff = useIsTrackSwitchedOff(track);

  return (
    <>
      {/* isEnabled is always 'true' on data tracks */}
      <Datum label="isEnabled" value="true" />
      {/* Only show isSwitchedOff and priority for remote data tracks */}
      {track.isSwitchedOff !== undefined && <Datum label="isSwitchedOff" value={isSwitchedOff} />}
      {track.priority !== undefined && <Datum label="Priority" value={track.priority} />}
      <DataTrack track={track} />
    </>
  );
};

export const DataTrackPublicationInfo: React.FC<{
  publication: LocalDataTrackPublication | RemoteDataTrackPublication;
}> = ({ publication }) => {
  const track = useTrack(publication) as LocalDataTrack | RemoteDataTrack | undefined;

  return (
    <StatsContainer>
      <Datum label="Kind" value={publication.kind} />
      <Datum label="isSubscribed" value={!!track} />
      {track && <DataTrackInfo track={track} />}
    </StatsContainer>
  );
};
