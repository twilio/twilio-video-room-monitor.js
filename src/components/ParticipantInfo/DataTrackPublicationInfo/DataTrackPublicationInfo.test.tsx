import React from 'react';
import { shallow } from 'enzyme';
import { DataTrackPublicationInfo, DataTrackInfo } from './DataTrackPublicationInfo';
import useTrack from '../../../hooks/useTrack/useTrack';

jest.mock('../../../hooks/useIsTrackSwitchedOff/useIsTrackSwitchedOff', () => () => true);
jest.mock('../../../hooks/useTrack/useTrack');

const mockUseTrack = useTrack as jest.Mock<any>;

mockUseTrack.mockImplementation(() => 'testTrack');

describe('the DataTrackInfo component', () => {
  it('should render correctly when a data track is present', () => {
    const wrapper = shallow(<DataTrackInfo track={{ isSwitchedOff: false, priority: 'high' } as any} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when isSwitchedOff and priority are undefined (for local data track)', () => {
    const wrapper = shallow(<DataTrackInfo track={{} as any} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('the DataTrackPublicationInfo component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<DataTrackPublicationInfo publication={{ trackName: 'testName' } as any} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render the DataTrackInfo component when a data track is not present', () => {
    mockUseTrack.mockImplementationOnce(() => null);
    const wrapper = shallow(<DataTrackPublicationInfo publication={{ trackName: 'testName' } as any} />);
    expect(wrapper.find(DataTrackInfo).exists()).toBe(false);
  });
});
