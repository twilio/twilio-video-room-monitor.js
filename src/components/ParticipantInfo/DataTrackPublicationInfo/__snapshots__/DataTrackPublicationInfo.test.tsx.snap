// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the DataTrackInfo component should render correctly when a data track is present 1`] = `
<Fragment>
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="isSwitchedOff"
    value={true}
  />
  <Memo(Datum)
    label="Priority"
    value="high"
  />
  <Memo(Tracks)
    track={
      Object {
        "isSwitchedOff": false,
        "priority": "high",
      }
    }
  />
</Fragment>
`;

exports[`the DataTrackInfo component should render correctly when isSwitchedOff and priority are undefined (for local data track) 1`] = `
<Fragment>
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Tracks)
    track={Object {}}
  />
</Fragment>
`;

exports[`the DataTrackPublicationInfo component should render correctly 1`] = `
<Memo(styled.div)>
  <Memo(Datum)
    label="Kind"
  />
  <Memo(Datum)
    label="isSubscribed"
    value={true}
  />
  <DataTrackInfo
    track="testTrack"
  />
</Memo(styled.div)>
`;
